torchtune/utils/_activations_offloading/offload_ctx_mgr.py

```diff
@@ -0,0 +1,260 @@
+    This is in contrast to maintaining the activation on GPU VRAM throughout the program.
+
+    This manager uses one additional CUDA stream to handle the communication between CUDA and CPU,
+    which is intended to overlap with the default computation stream to improve runtime. We designed
+    synchronization with a few heuristics for optimizing the tradeoff between runtime vs memory usage.
+
+    max_fwd_stash_size: int = 5. The maximum size of the forward stash, or the maximum number of
+        consecutive activations to keep alive during the forward pass. This number must be at
+        least 1. Keeping alive more activations will potentially allow more overlap between the
+        communication and compute streams at the cost of increasing memory usage. Keeping alive
+        fewer activations will conserve memory, but may cause poor overlap between the streams,
+        increasing runtime. Default: 5.
+
+    min_offload_size: int = 1024. The minimum number of bytes a Tensor must be in order to qualify
+        for offloading. If the tensor is too small, we do not want to waste bandwidth and resources
+        moving it to CPU and back. Default: 1024 bytes.
+
+    use_pin_memory: bool = True. Whether or not the offloaded Tensor will be placed in pinned
+        memory on the CPU. Pinned memory allows the Tensor to be moved back onto GPU more quickly
+        but is a limited resource. Default: True.
+    """
+
+    def __init__(
+        self,
+        max_fwd_stash_size: int = 5,
+        min_offload_size: int = 1024,
+        use_pin_memory: bool = True,
+    ) -> None:
+        self.min_tensor_size_bytes = (
+            min_offload_size  # we don't want to bother with small tensors
+        )
+        self.tracker = (
+            {}
+        )  # tensor_id => (new_tensor, if_modified)  ---> track what saved/offloaded tensors are where
+        self.tensor_id: int = 0
+        self.is_first_forward_call = True
+        self.is_first_backward_call = True
+        self.is_first_forward_pass = True
+
+        # managing cpu memory
+        self.use_pin_memory: bool = use_pin_memory
+        self.virtual_memory_safe_pct = (
+            0.60  # we should not exceed this percentage of memory
+        )
+
+        # for streaming
+        self.s0 = torch.cuda.default_stream()  # comp stream
+        self.s1 = torch.cuda.Stream()  # comms stream
+        self.fwd_stash = {}  # tensor_id => (activation, ev1)
+        if max_fwd_stash_size < 1:
+            raise ValueError(
+                f"max_fwd_stash_size should be at least 1 but is {max_fwd_stash_size}"
+            )
+        self.max_fwd_stash_size = max_fwd_stash_size
+        self.bwd_tensor_stash = {}  # tensor_id => activation
+        self.bwd_ev_stash = {}  # tensor_id => ev0
+        self.curr_graph_id = None
+        self.curr_autograd_node = None
+
+        # platform util functions
+        def verify_sufficient_virtual_memory():
```

Comment:
is there a motivation to define so many functions within __init__? i.e. why not do it as a method of the class or as another function outside of the class?

Reply:
It's cuz everything gets accessed from the __init__, so everything got defined here for linearity. I don't have any real preference though. Is there a reason to move them out? If so, where would be best? (In the class but as a class method or completely out of the class.)

Reply:
Oh actually, I know why everything's in the init lol. It's cuz `pack` can't take in any variables other than the tensor --> so we can't make that a class method that takes in `self`. And technically we could move it to the global space and have `OffloadActivations` write a bunch of metadata to global, but that doesn't seem better.

Reply:
we _could_ go and delete all the self.'s though lol